File: domain/src/test/java/org/cryptomator/domain/usecases/cloud/UploadFileTest.java
package org.cryptomator.domain.usecases.cloud;

import android.content.Context;

import org.cryptomator.domain.CloudFile;
import org.cryptomator.domain.CloudFolder;
import org.cryptomator.domain.exception.BackendException;
import org.cryptomator.domain.repository.CloudContentRepository;
import org.cryptomator.domain.usecases.ProgressAware;
import org.cryptomator.util.Optional;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.ValueSource;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.List;

import static java.util.Arrays.fill;
import static java.util.Collections.singletonList;
import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.mockito.Matchers.any;
import static org.mockito.Matchers.eq;
import static org.mockito.Matchers.same;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class UploadFileTest {

	private final Context context = mock(Context.class);
	private final CloudContentRepository cloudContentRepository = mock(CloudContentRepository.class);
	private final CloudFolder parent = mock(CloudFolder.class);
	private final CloudFile targetFile = mock(CloudFile.class);
	private final CloudFile resultFile = mock(CloudFile.class);
	private final String fileName = "fileName";
	private final ProgressAware<UploadState> progressAware = mock(ProgressAware.class);

	@ParameterizedTest
	@ValueSource(booleans = {true, false})
	public void testInvocationWithFileSizeDelegatesToCloudContentRepository(Boolean replacing) throws BackendException {
		long fileSize = 1337;
		DataSource dataSource = dataSourceWithBytes(0, fileSize, Optional.of(fileSize));
		UploadFiles inTest = testCandidate(dataSource, replacing);
		when(cloudContentRepository.file(parent, fileName, Optional.of(fileSize))).thenReturn(targetFile);
		when(cloudContentRepository.write(same(targetFile), any(DataSource.class), same(progressAware), eq(replacing), eq(fileSize))).thenReturn(resultFile);

		List<CloudFile> result = inTest.execute(progressAware);

		assertThat(result.size(), is(1));
		assertThat(result.get(0), is(resultFile));
	}

	@ParameterizedTest
	@ValueSource(booleans = {true, false})
	public void testInvocationWithoutFileSizeDelegatesToCloudContentRepository(Boolean replacing) throws BackendException, IOException {
		long fileSize = 8893;
		try (DataSource dataSource = dataSourceWithBytes(85, fileSize, Optional.empty())) {
			UploadFiles inTest = testCandidate(dataSource, replacing);
			when(cloudContentRepository.file(parent, fileName, Optional.of(fileSize))).thenReturn(targetFile);
			DataSourceCapturingAnswer capturedStreamData = new DataSourceCapturingAnswer(resultFile, 1);
			when(cloudContentRepository.write(same(targetFile), any(DataSource.class), same(progressAware), eq(replacing), eq(fileSize))).thenAnswer(capturedStreamData);

			List<CloudFile> result = inTest.execute(progressAware);

			assertThat(result.size(), is(1));
			assertThat(result.get(0), is(resultFile));
			assertThat(capturedStreamData.toByteArray(), is(bytes(85, fileSize)));
		}
	}

	private DataSource dataSourceWithBytes(int value, long amount, final Optional<Long> size) {
		if (amount > Integer.MAX_VALUE) {
			throw new IllegalStateException("Can not use values > Integer.MAX_VALUE");
		}
		final byte[] bytes = bytes(value, (int) amount);
		return new DataSource() {

			@Override
			public Optional<Long> size(Context context) {
				return size;
			}

			@Override
			public InputStream open(Context context) throws IOException {
				return new ByteArrayInputStream(bytes);
			}

			@Override
			public DataSource decorate(DataSource delegate) {
				return delegate;
			}

			@Override
			public void close() throws IOException {
				// do nothing
			}
		};
	}

	private byte[] bytes(int value, long amount) {
		if (amount > Integer.MAX_VALUE) {
			throw new IllegalStateException("Can not use values > Integer.MAX_VALUE");
		}
		byte[] data = new byte[(int) amount];
		fill(data, (byte) value);
		return data;
	}

	private UploadFiles testCandidate(DataSource dataSource, Boolean replacing) {
		return new UploadFiles( //
				context, //
				cloudContentRepository, //
				parent, //
				singletonList(new UploadFile.Builder() //
						.withFileName(fileName) //
						.withDataSource(dataSource) //
						.thatIsReplacing(replacing) //
						.build()));
	}

}

File: domain/src/test/java/org/cryptomator/domain/usecases/cloud/UploadFileTest.kt
package org.cryptomator.domain.usecases.cloud

import android.content.Context
import org.cryptomator.domain.Cloud
import org.cryptomator.domain.CloudFile
import org.cryptomator.domain.CloudFolder
import org.cryptomator.domain.CloudNode
import org.cryptomator.domain.exception.BackendException
import org.cryptomator.domain.repository.CloudContentRepository
import org.cryptomator.domain.usecases.ProgressAware
import org.cryptomator.util.Optional
import org.hamcrest.CoreMatchers
import org.hamcrest.MatcherAssert
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.ValueSource
import org.mockito.Mockito
import org.mockito.kotlin.eq
import org.mockito.kotlin.mock
import org.mockito.kotlin.same
import org.mockito.kotlin.verify
import org.mockito.kotlin.whenever
import java.io.ByteArrayInputStream
import java.io.IOException
import java.io.InputStream
import java.util.Arrays
import java.util.Date

class UploadFileTest {

	private val context: Context = mock()
	private var cloudContentRepository: CloudContentRepository<Cloud, CloudNode, CloudFolder, CloudFile> = mock()
	private val parent: CloudFolder = mock()
	private val targetFile: CloudFile = mock()
	private val resultFile: CloudFile = mock()

	private val progressAware: ProgressAware<UploadState> = mock()

	private val fileName = "fileName"

	private fun <T> any(type: Class<T>): T = Mockito.any(type)

	@ParameterizedTest
	@ValueSource(booleans = [true, false])
	@Throws(BackendException::class)
	fun testInvocationWithFileSizeDelegatesToCloudContentRepository(replacing: Boolean) {
		val fileSize: Long = 1337
		val dataSource = dataSourceWithBytes(0, fileSize, fileSize)
		val inTest = testCandidate(dataSource, replacing)

		whenever(cloudContentRepository.file(parent, fileName, fileSize)).thenReturn(targetFile)
		whenever(
			cloudContentRepository.write(
				same(targetFile),
				any(DataSource::class.java),
				same(progressAware),
				eq(replacing),
				eq(fileSize)
			)
		).thenReturn(resultFile)

		val result = inTest.execute(progressAware)

		MatcherAssert.assertThat(result.size, CoreMatchers.`is`(1))
		MatcherAssert.assertThat(result[0], CoreMatchers.`is`(resultFile))
	}

	@ParameterizedTest
	@ValueSource(booleans = [true, false])
	@Throws(BackendException::class, IOException::class)
	fun testInvocationWithoutFileSizeDelegatesToCloudContentRepository(replacing: Boolean) {
		val fileSize: Long = 8893
		dataSourceWithBytes(85, fileSize, null).use { dataSource ->
			val inTest = testCandidate(dataSource, replacing)
			whenever(cloudContentRepository.file(parent, fileName, fileSize)).thenReturn(targetFile)
			val capturedStreamData = DataSourceCapturingAnswer<Any?>(resultFile, 1)
			whenever(
				cloudContentRepository.write(
					same(targetFile),
					any(DataSource::class.java),
					same(progressAware),
					eq(replacing),
					eq(fileSize)
				)
			).thenAnswer(capturedStreamData)

			val result = inTest.execute(progressAware)

			MatcherAssert.assertThat(result.size, CoreMatchers.`is`(1))
			MatcherAssert.assertThat(result[0], CoreMatchers.`is`(resultFile))
			MatcherAssert.assertThat(capturedStreamData.toByteArray(), CoreMatchers.`is`(bytes(85, fileSize)))
		}
	}

	private fun dataSourceWithBytes(value: Int, amount: Long, size: Long?): DataSource {
		check(amount <= Int.MAX_VALUE) { "Can not use values > Integer.MAX_VALUE" }
		val bytes = bytes(value, amount)
		return object : DataSource {
			override fun size(context: Context): Long? {
				return size
			}

			@Throws(IOException::class)
			override fun open(context: Context): InputStream {
				return ByteArrayInputStream(bytes)
			}

			override fun decorate(delegate: DataSource): DataSource {
				return delegate
			}

			override fun modifiedDate(context: Context): Optional<Date> {
				return Optional.of(Date())
			}

			@Throws(IOException::class)
			override fun close() {
				// do nothing
			}
		}
	}

	private fun bytes(value: Int, amount: Long): ByteArray {
		check(amount <= Int.MAX_VALUE) { "Can not use values > Integer.MAX_VALUE" }
		val data = ByteArray(amount.toInt())
		Arrays.fill(data, value.toByte())
		return data
	}

	@Test
	@DisplayName("Completed files are skipped during upload")
	@Throws(BackendException::class)
	fun testCompletedFilesAreSkipped() {
		val fileSize: Long = 1337
		val dataSource1 = dataSourceWithBytes(0, fileSize, fileSize)
		val dataSource2 = dataSourceWithBytes(1, fileSize, fileSize)
		val targetFile2: CloudFile = mock()
		val resultFile2: CloudFile = mock()

		val files = listOf(
			UploadFile.Builder()
				.withFileName("file1.txt")
				.withDataSource(dataSource1)
				.thatIsReplacing(false)
				.build(),
			UploadFile.Builder()
				.withFileName("file2.txt")
				.withDataSource(dataSource2)
				.thatIsReplacing(false)
				.build()
		)

		val inTest = UploadFiles(context, cloudContentRepository, parent, files)
		inTest.setCompletedFiles(setOf("file1.txt"))

		whenever(cloudContentRepository.file(parent, "file2.txt", fileSize)).thenReturn(targetFile2)
		whenever(
			cloudContentRepository.write(
				same(targetFile2),
				any(DataSource::class.java),
				same(progressAware),
				eq(false),
				eq(fileSize)
			)
		).thenReturn(resultFile2)

		val result = inTest.execute(progressAware)

		MatcherAssert.assertThat(result.size, CoreMatchers.`is`(1))
		MatcherAssert.assertThat(result[0], CoreMatchers.`is`(resultFile2))
	}

	@Test
	@DisplayName("FileUploadedCallback is called after each file upload")
	@Throws(BackendException::class)
	fun testFileUploadedCallbackIsCalled() {
		val fileSize: Long = 100
		val dataSource = dataSourceWithBytes(0, fileSize, fileSize)
		val callback: FileUploadedCallback = mock()

		val files = listOf(
			UploadFile.Builder()
				.withFileName(fileName)
				.withDataSource(dataSource)
				.thatIsReplacing(false)
				.build()
		)

		val inTest = UploadFiles(context, cloudContentRepository, parent, files)
		inTest.setFileUploadedCallback(callback)

		whenever(cloudContentRepository.file(parent, fileName, fileSize)).thenReturn(targetFile)
		whenever(
			cloudContentRepository.write(
				same(targetFile),
				any(DataSource::class.java),
				same(progressAware),
				eq(false),
				eq(fileSize)
			)
		).thenReturn(resultFile)

		inTest.execute(progressAware)

		verify(callback).onFileUploaded(fileName)
	}

	@Test
	@DisplayName("All completed files skipped returns empty list")
	@Throws(BackendException::class)
	fun testAllFilesCompletedReturnsEmpty() {
		val fileSize: Long = 100
		val dataSource = dataSourceWithBytes(0, fileSize, fileSize)

		val files = listOf(
			UploadFile.Builder()
				.withFileName("a.txt")
				.withDataSource(dataSource)
				.thatIsReplacing(false)
				.build(),
			UploadFile.Builder()
				.withFileName("b.txt")
				.withDataSource(dataSource)
				.thatIsReplacing(false)
				.build()
		)

		val inTest = UploadFiles(context, cloudContentRepository, parent, files)
		inTest.setCompletedFiles(setOf("a.txt", "b.txt"))

		val result = inTest.execute(progressAware)

		MatcherAssert.assertThat(result.size, CoreMatchers.`is`(0))
	}

	private fun testCandidate(dataSource: DataSource, replacing: Boolean): UploadFiles {
		return UploadFiles( //
			context,  //
			cloudContentRepository,  //
			parent,  //
			listOf(
				UploadFile.Builder() //
					.withFileName(fileName) //
					.withDataSource(dataSource) //
					.thatIsReplacing(replacing) //
					.build()
			)
		)
	}
}
